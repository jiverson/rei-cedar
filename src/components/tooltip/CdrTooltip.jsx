import style from './styles/CdrTooltip.scss';
import CdrPopup from '../popup/CdrPopup';
import propValidator from '../../utils/propValidator';

export default {
  name: 'CdrTooltip',
  components: {
    CdrPopup,
  },
  props: {
    position: {
      type: String,
      required: false,
      default: 'up',
      validator: (value) => propValidator(
        value,
        ['up', 'down', 'left', 'right'],
      ),
    },
    autoPosition: {
      type: Boolean,
      default: true,
    },
  },
  data() {
    return {
      style,
      open: false,
      openHandler: undefined,
      closeHandler: undefined,
      timeout: undefined,
    };
  },
  mounted() {
    this.addHandlers();
  },
  destroyed() {
    // TODO: no need to do this since the element listened to gets destroyted with this too?
    // const triggerElement = this.$refs.trigger.children[0];
    // if (triggerElement) {
    //   triggerElement.removeEventListener('mouseover', this.openHandler);
    //   triggerElement.removeEventListener('focus', this.openHandler);
    //
    //
    //   triggerElement.removeEventListener('mouseleave', this.closeHandler);
    //   triggerElement.removeEventListener('blur', this.closeHandler);
    // }
  },
  methods: {
    openTooltip() {
      if (this.timeout) clearTimeout(this.timeout);
      this.open = true;
    },
    closeTooltip() {
      this.timeout = setTimeout(() => {
        this.open = false;
      }, 250);
    },
    addHandlers() {
      this.openHandler = this.openTooltip.bind(this);
      this.closeHandler = this.closeTooltip.bind(this);

      const triggerElement = this.$refs.trigger.children[0];
      const popupElement = this.$refs.popup.$el;
      if (triggerElement) {
        triggerElement.addEventListener('mouseover', this.openHandler);
        triggerElement.addEventListener('focus', this.openHandler);

        triggerElement.addEventListener('mouseleave', this.closeHandler);
        triggerElement.addEventListener('blur', this.closeHandler);

        popupElement.addEventListener('mouseover', this.openHandler);
        popupElement.addEventListener('mouseleave', this.closeHandler);
      }
    },
  },
  render() {
    // TODO: aria-hidden/opened/expanded logic for tooltip?
    // aria-live when content gets displayed?
    return (
      <div
        class={this.style['cdr-tooltip--wrapper']}
      >
        <div ref="trigger">
          { this.$slots.trigger }
        </div>
        <cdr-popup
          class={this.style['cdr-tooltip']}
          role="tooltip"
          ref="popup"
          position={ this.position }
          autoPosition={ this.autoPosition }
          opened={ this.open }
        >
          { this.$slots.default }
        </cdr-popup>
      </div>
    );
  },
};
