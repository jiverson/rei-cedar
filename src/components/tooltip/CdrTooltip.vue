\import clsx from 'clsx';
import style from './styles/CdrTooltip.scss';
import CdrPopup from '../popup/CdrPopup';
import propValidator from '../../utils/propValidator';

export default {
  name: 'CdrTooltip',
  components: {
    CdrPopup,
  },
  props: {
    position: {
      type: String,
      required: false,
      default: 'top',
      validator: (value) => propValidator(
        value,
        ['top', 'bottom', 'left', 'right'],
      ),
    },
    autoPosition: {
      type: Boolean,
      default: true,
    },
    id: {
      type: String,
      required: true,
    },
    contentClass: {
      type: String,
      required: false,
    },
    open: {
      type: Boolean,
      default: false,
      required: false,
    },
  },
  data() {
    return {
      style,
      isOpen: false,
      openHandler: undefined,
      closeHandler: undefined,
      timeout: undefined,
    };
  },
  watch: {
    open() {
      if (this.open) {
        this.openTooltip();
      } else {
        this.closeTooltip();
      }
    },
  },
  mounted() {
    this.addHandlers();
    const trigger = this.$refs.trigger.children[0];
    if (trigger) trigger.setAttribute('aria-describedby', this.id);
  },
  methods: {
    openTooltip(e) {
      if (this.timeout) clearTimeout(this.timeout);
      this.isOpen = true;
      this.$emit('opened', e);
    },
    closeTooltip(e) {
      this.timeout = setTimeout(() => {
        this.isOpen = false;
        this.$emit('closed', e);
      }, 250);
    },
    addHandlers() {
      this.openHandler = this.openTooltip.bind(this);
      this.closeHandler = this.closeTooltip.bind(this);

      const triggerElement = this.$refs.trigger.children[0];
      const popupElement = this.$refs.popup.$el;
      if (triggerElement) {
        triggerElement.addEventListener('mouseover', this.openHandler);
        triggerElement.addEventListener('focus', this.openHandler);

        triggerElement.addEventListener('mouseleave', this.closeHandler);
        triggerElement.addEventListener('blur', this.closeHandler);

        popupElement.addEventListener('mouseover', this.openHandler);
        popupElement.addEventListener('mouseleave', this.closeHandler);
      }
    },
  },
  render() {
    return (
      <div class={clsx(
        this.style['cdr-tooltip--wrapper'],
        this.$slots.trigger ? this.style['cdr-tooltip--position'] : '',
      )}>
        <div ref="trigger">
          { this.$slots.trigger }
        </div>
        <cdr-popup
          class={this.style['cdr-tooltip']}
          contentClass={ this.contentClass }
          role="tooltip"
          ref="popup"
          position={ this.position }
          autoPosition={ this.autoPosition }
          opened={ this.isOpen }
          id={this.id}
          onClosed={ this.closeTooltip }
        >
          { this.$slots.default }
        </cdr-popup>
      </div>
    );
  },
};
