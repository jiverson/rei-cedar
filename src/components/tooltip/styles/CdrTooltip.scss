@import '../../../css/settings/index.scss';
@import url('@rei/cedar/dist/style/cdr-popup.css');

// TODO: delete these once tokens is updated
$cdr-color-border-tooltip-default: #f9f8f6;
$cdr-color-background-tooltip-default: #373734;
$cdr-color-text-tooltip-default: #f9f8f6;

.cdr-tooltip--wrapper {
  position: relative;
  width: max-content;
  height: max-content;

  .cdr-popup__content {
    background: $cdr-color-background-tooltip-default;
    color: $cdr-color-text-tooltip-default;
    border: 1px solid $cdr-color-border-tooltip-default;
    border-radius: $cdr-radius-softer;
    @include cdr-text-utility-strong-100;
    padding: $cdr-space-inset-three-quarter-x-squish;
  }

  // set arrow color on popup
  .cdr-popup--down .cdr-popup__arrow {
    &::before {
      border-bottom-color: $cdr-color-border-tooltip-default;
    }

    &::after {
      border-bottom-color: $cdr-color-background-tooltip-default;
    }
  }

  .cdr-popup--up .cdr-popup__arrow {
    &::before {
      border-top-color: $cdr-color-border-tooltip-default;
    }

    &::after {
      border-top-color: $cdr-color-background-tooltip-default;
    }
  }

  .cdr-popup--right .cdr-popup__arrow {
    &::before {
      border-right-color: $cdr-color-border-tooltip-default;
    }

    &::after {
      border-right-color: $cdr-color-background-tooltip-default;
    }
  }

  .cdr-popup--left .cdr-popup__arrow {
    &::before {
      border-left-color: $cdr-color-border-tooltip-default;
    }

    &::after {
      border-left-color: $cdr-color-background-tooltip-default;
    }
  }
}
