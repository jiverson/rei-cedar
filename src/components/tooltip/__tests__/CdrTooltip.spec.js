import { shallowMount } from '@vue/test-utils';
import CdrTooltip from 'componentdir/tooltip/CdrTooltip';

describe('CdrTooltip', () => {
  it('matches snapshot', () => {
    const wrapper = shallowMount(CdrTooltip);
    expect(wrapper.element).toMatchSnapshot();
  });
});
