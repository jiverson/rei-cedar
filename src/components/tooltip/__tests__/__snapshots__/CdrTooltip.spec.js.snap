// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CdrTooltip matches snapshot 1`] = `
<div
  class="cdr-tooltip--wrapper"
>
  <div />
  <cdr-popup-stub
    autoposition="true"
    position="up"
    role="tooltip"
  />
</div>
`;
